torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,148 @@
+import itertools
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    buffers outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" forward once to initialize all parameters in case
+    that you need to access the weights before the actual real ``forward`` pass.
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
+    dummy forward call in the following scenarios.
+
+
+    >>> lazy_module = LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
+    >>> type(lazy_module)
+
+    * Access an :class:`UnitializedParameter` data or properties
+    >>> lazy_module.weight.shape
+
+    * Use an Optimizer that access Parameters data in their constructor 
+      (Currently only :class:`torch.optim.AdaGrad`) other optimizers work without
+      explicitly initializing the module
+
+    >>> optim = torch.optim.AdaGrad(lazy_module.parameters())
+
+    * Use :class:`torch.nn.DataParallel` or :class:`torch.nn.DistributedDataParallel`
+    >>> module_dp = torch.nn.DataParallel(lazy_module)
+
+    * Module reparametrization using :func:`torch.nn.utils.weight_norm` or :func:`torch.nn.utils.spectral_norm`
+    >>> torch.nn.utils.weight_norm(module)
```

Comment:
Also, where's an example of non-default weight initialization? It can be said to be implicit in "access data or properties", but probably makes sense to also mention it explicitly.